### PR TITLE
Wip wjw freebsd tests

### PR DESCRIPTION
**Pull request by wjwithagen**
If everything for Ceph compiles under FreeBSD
And src/test/run-cli-tests runs to succes
Start compiling the tests

### CODE REVIEW

**Comment by tchaikov** [comment]
note for myself:

- [ ] FreeBSD: Scripts to configure/build/compile …	9670d0d
- [ ] FreeBSD: Add dependancies for building under FreeBSD …	1b483f5
- [ ] FreeBSD: fix confiure items that are needed under FreeBSD …	c183dc9
- [ ] FreeBSD/autobuild: extra whitespace changed relative to head …	  9a73bd0
- [ ] FreeBSD/malloc: Mallinfo is not available in FreeBSD …	ea51474
- [ ] FreeBSD/autobuild: Still one more -ldl that needs to be Linux-ized …	6cdba97
- [ ] FreeBSD/pthread: The non-POSIX routines are in a different include file …	ea95717
- [ ] FreeBSD/features: This file is not on FreebSD. …	ae303fe
- [ ] FreeBSD/Bluestore: exclude Bluestore types from being generated. …	  854f39e
- [ ] FreeBSD/xattr: Use a different include, and fix CONSTS …	e89e50e
- [ ] FreeBSD/autobuild: fix leading spaces for defines …	df844c9
- [ ] FreeBSD/autobuild: Build without Rockdb for the time being …	a2a92e1
- [ ] FreeBSD/build: The prerequisite is that Ceph needs /bin/bash to work. …	482409a
- [ ] common/test_time.cc: Clang and GCC have different opinions on us of c… …	c644500
- [ ] common/Graylog.cc: GCC and Clang handle casting differently. …	  fdc0a9b
- [ ] FreeBSD/tests/run-cli-test: FreeBSD does not have RBD, skip tests …	e5fd5f4
- [ ] FreeBSD/POSIX: tests in src/test/cli used grep -P …	f3c22a2
- [ ] FreeBSD/sockets: Socket constants and functions are defined in <sys/s… …	f43e1cc
- [ ] FreeBSD/execinfo: FreeBSD libexecinfo needs to be linked explicitly. …	f7434c6
- [ ] FreeBSD/unittest_on_exit: Anonymous mmap does not like fd 0 as fileha… …	2a9099b
- [ ] FreeBSD/init: The tests for the rc.d inti system are not present …	30dd424
- [ ] FreeBSD/ceph-disk.sh: missing blkid on FreeBSD, zo skip these tests …	dc2e0c2
- [ ] FreeBSD/ceph-disk.sh: stat -c '%T' does not give fstype …	722a501
- [ ] FreeBSD/Clang: buffer::ptr enc[k + m] declaration only works when k &… …	3375ca5
- [ ] FreeBSD/ceph_disk/main.py: /proc/mount lives in /compat/linux. …	b8cd9cd
- [ ] FreeBSD/mutex_debug.h: Manually create the debug mutex …	561f615
- [ ] FreeBSD/ceph-helpers.sh test_pidfile.sh: fix more use of stat -C %T …	062aeb1
- [x] FreeBSD/pidfile.cc: struct flock is different on FreeBSD. …	82d6baa
- [ ] FreeBSD/FileJournal.cc NO verifing cache functionality on FreeBSD …	02c9809
- [ ] FreeBSD/stop.sh Only unmap rbd images if we have rbd available …	8a8a97d
- [ ] FreeBSD//cross_process_sem.cc: anonymous mmap needs -1 as fd-handle …	95ccbdf
- [x] FreeBSD/vstart.sh: Freebsd does not have btrfs. …	9a18d49
- [ ] FreeBSD/chain_xattr.cc: attributes are not return in specific order. …	ec2a0ea
- [x] test/confutils.cc: Report that the atexit routine is really excuted. …	eba3952
- [x] FreeBSD/ceph-disk.sh: Test some debugging code …unittest_addrs

File: src/test/test_pool_create.sh [hunk 1]
```diff
@@ -34,15 +35,27 @@ function TEST_pool_create() {
...
+    ceph osd getcrushmap -o  $dir/map1
+    crushtool -d $dir/map1 -o $dir/map1.txt
 
     local minsize=0
     local maxsize=1
-    sed -i '/# end crush map/i\rule '$rulename' {\n ruleset \'$var'\n type replicated\n min_size \'$minsize'\n max_size \'$maxsize'\n step take default\n step choose firstn 0 type osd\n step emit\n }\n' map1.txt
-    crushtool  -c map1.txt -o  map1.bin
-    ceph osd setcrushmap -i map1.bin
-    ceph osd  pool create  $poolname 200 $rulename 2>rev
+	cat >> $dir/map1.txt << _EOF_RULE
```
Comment (songbaisen) [inline src/test/test_pool_create.sh]: Would you please remove the space before the cat to make here neat.

File: src/test/test_pool_create.sh [hunk 2]
```diff
@@ -34,15 +35,27 @@ function TEST_pool_create() {
...
+    rule $rulename {
+	    ruleset $var
+	    type replicated
+	    min_size $minsize
+	    max_size $maxsize
+	    step take default
+	    step choose firstn 0 type osd
+	    step emit
+	}
+_EOF_RULE
```
Comment (songbaisen) [inline src/test/test_pool_create.sh]: Thank you for your restructure it  here. LGTM!

Comment (wjwithagen) [inline src/test/test_pool_create.sh]: @songbaisen 
Fixed the TABS, but also found another reference to a output file that was going to the current directory (rev).
Also prefixed that with $dir to get it into testdir.

Comment (songbaisen) [inline src/test/test_pool_create.sh]: @wjwithagen  cool!Bingo!

**Comment by tchaikov** [comment]
@wjwthagen could you split this pull request into smaller pieces? it could be difficult for you to keep up with the master, and difficult for the reviewer to digest these commits. not to mention some of them are only for debugging purpose, and not supposed to be merged, IIUC.

for example, you could put the changes which applease clang into an individual PR?

**Comment by wjwithagen** [comment]
On 21-3-2016 16:57, Kefu Chai wrote:
> @wjwthagen could you split this pull request into smaller pieces? it
> could be difficult for you to keep up with the master, and difficult for
> the reviewer to digest these commits. not to mention some of them are
> only for debugging purpose, and not supposed to be merged, IIUC.
> 
> for example, you could put the changes which applease clang into an
> individual PR?

Hi Kefu,

I was afraid that this was going to be the question at some point ....
8D.

But the pull grew thus big, that I can understand the hesitation to take
it as one big clunk.

That is why it is actually 3 pulls:
	wip-wjw-freebsd-autobuild #7452
	wip-wjw-freebsd-compile   #7515
	wip-wjw-freebsd-tests     #7573

And the wip-wjw-freebsd-autobuild pull was exactly to be intended what
you are asking for.

But that feeling has blurred, because during the testing, I still needed
to augment autobuild things. Either because of required changes, or just
because I changed my view on certain things.

Perhaps I should just reorganise, fixup and squash the set of commits to
be more of the original intentions.

As far as changes for debugging, I try not to commit changes that are
not releavant for running the tests. Currently the tests are strongly
biased tot he fact that the test should always complete oke.

Which in my certainly is not the case. :( So I'm adding code to make
sure that more porting/testing/debugging work will be easier because the
tests are more consistently leaving post-mortum data behind that can be
looked at.
EG. in my case it is rather clumsy that the logs from OSDs and MONs are
dumped into test-suites.log. It is much better to leave the testdir as
is, and look at it there.

But that is a process of growing insight, hence the cummulation of
several commits. And yes they deserve cleanup.

But I would already be helped a lot if I "can get ride" of the first
part of my pull: the autobuild stuff, and the compile stuff.

So I can do that by reorganising with:
	git rebase -i 5c6e545^

And group the commits per type, and then perhaps fixup or squash?

Is that something that would work?

--WjW

**Comment by tchaikov** [comment]
thanks @wjwithagen i will try to focus #7515 first. and some of your commits are showing up in multiple pull requests of yours. for example, 14f0ef0e1ba4680e395f043a71966b12b7e9a0bb can be found in #7452 and #7515. it's quite confusing to look at them, imho. TBH, i found you are stashing your changes into your PRs/remote branches, but if the PRs are intended for prompt review. we'd better keep them in shape and updated.

and again, the smaller the easier.

**Comment by wjwithagen** [comment]
On 22-3-2016 05:06, Kefu Chai wrote:
> thanks @wjwithagen <https://github.com/wjwithagen> i will try to focus
> #7515 <https://github.com/ceph/ceph/pull/7515> first. and some of your
> commits are showing up in multiple pull requests of yours. for example,
> 14f0ef0
> <https://github.com/ceph/ceph/commit/14f0ef0e1ba4680e395f043a71966b12b7e9a0bb>
> can be found in #7452 <https://github.com/ceph/ceph/pull/7452> and #7515
> <https://github.com/ceph/ceph/pull/7515>. it's quite confusing to look
> at them, imho. TBH, i found you are stashing your changes into your
> PRs/remote branches, but if the PRs are intended for prompt review. we'd
> better keep them in shape and updated.

Yes,

I discussed this with Sage on how to make progress and thus I ended up
with a PULL on a PULL on a PULL. But is has become way to big.

So the last PULL contains every thing that is also in the first.
And the idea is that the first: #7452 addresses the autobuild
infrastructure. So perhaps we can focus on that first.
Then #7515 should be all about actual changes to the c/C++ code to get
everything compiled (including the testset, without BlueStore and RBD)
After which the testset is addressed in this pull (#7573)

So if we can get #7452 out of the way first, that'll reduce a basic
chunk of the pull. However I see that there are still compiler commits
in the autobuild part. This rebasing does not always do what I want it
to do, and even after a push things look different than they did here
local. :(

Perhaps making a new PULL with all autobuild changes per file squashed
would work better?

**Comment by wjwithagen** [comment]
On 22-3-2016 11:31, Willem Jan Withagen wrote:
> On 22-3-2016 05:06, Kefu Chai wrote:
>> thanks @wjwithagen <https://github.com/wjwithagen> i will try to focus
>> #7515 <https://github.com/ceph/ceph/pull/7515> first. and some of your
>> commits are showing up in multiple pull requests of yours. for example,
>> 14f0ef0
>> <https://github.com/ceph/ceph/commit/14f0ef0e1ba4680e395f043a71966b12b7e9a0bb>
>> can be found in #7452 <https://github.com/ceph/ceph/pull/7452> and #7515
>> <https://github.com/ceph/ceph/pull/7515>. it's quite confusing to look
>> at them, imho. TBH, i found you are stashing your changes into your
>> PRs/remote branches, but if the PRs are intended for prompt review. we'd
>> better keep them in shape and updated.

@tchaikov, ping...

> Yes,
> 
> I discussed this with Sage on how to make progress and thus I ended up
> with a PULL on a PULL on a PULL. But is has become way to big.
> 
> So the last PULL contains every thing that is also in the first.
> And the idea is that the first: #7452 addresses the autobuild
> infrastructure. So perhaps we can focus on that first.
> Then #7515 should be all about actual changes to the c/C++ code to get
> everything compiled (including the testset, without BlueStore and RBD)
> After which the testset is addressed in this pull (#7573)
> 
> So if we can get #7452 out of the way first, that'll reduce a basic
> chunk of the pull. However I see that there are still compiler commits
> in the autobuild part. This rebasing does not always do what I want it
> to do, and even after a push things look different than they did here
> local. :(
> 
> Perhaps making a new PULL with all autobuild changes per file squashed
> would work better?
> 
> 
>

**Comment by tchaikov** [comment]
>  Perhaps making a new PULL with all autobuild changes per file squashed 
>  would work better?

yes.

**Comment by wjwithagen** [comment]
@tchaikov 
Right, got to put some work in it.
I'll ping you with a new PULL

**Comment by tchaikov** [comment]
@wjwthagen

not necessarily "per file squashed", but ideally, it would be better if each commit is logically independent.

**Comment by wjwithagen** [comment]
@tchaikov 
I'm just going to checkout a fresh master, run a diff against it.
And commit the diffs that have to do with the autobuild.
So they are all fresh commit, and neatly summarised. 
Just as easy.

**Comment by wjwithagen** [comment]
@tchaikov 
I've you have the time, please look at #8298 
That will generate the Makefiles and start compiling, but then the compiler bombs due to
erroneous/missing includes.
That will be the next set of commits in a seperate pull, but lets do this one first.

**Comment by tchaikov** [comment]
@wjwithagen i am closing it as you are posting them in smaller and manageable batches.

**Comment by wjwithagen** [comment]
@tchaikov 

Oke,....

The first 2 were really 2 large steps to get to actual building.
I'm fine with smaller steps from now on.

Although it is a delicate balance between the time waiting for QA and the amount that you'd like to submit. Running small batches creates a long turnnover time. Unless you want to pipeline by taking small commit batches in parallel through QA?

The trick will be not to disturb the QA stuff for linux, so things go in smoothly

**Pull request closed without merging**